### DISCUSSION TITLE
insert_text and overlapping text! .

### DISCUSSION THREAD

**Discussion by Muhammadraafat1** [pymupdf/PyMuPDF#3521] [Looking for help] [answered]
Hello, I have a problem here:
we should replace "610 E Griffin Pkwy Suite D Mission, TX" to "310 W Palmirs Blvd Suite F Antonio, NM", but I don't know why the result not accurate! I think I wrote everything right, Can anyone help? 

```
rep = {
    "610 E Griffin Pkwy Suite D Mission, TX": "910 W Palmirs Blvd Suite F Antonio, NM",
    "78572": "12345",
    "TX": "NM"
}

# Open the document
doc = fitz.open("original.pdf")

def is_overlapping(rect1, rect2):
    return not (rect1.x1 <= rect2.x0 or rect1.x0 >= rect2.x1 or rect1.y1 <= rect2.y0 or rect1.y0 >= rect2.y1)

replacement_positions = []
for page_num, page in enumerate(doc):
    fitz.TOOLS.set_small_glyph_heights(True)
    for word, replace in rep.items():
        wlist = page.search_for(word)
        for rect in wlist:
            # Check if this rect overlaps with an existing rect for the same word
            for existing_page_num, existing_rect, existing_word, existing_replace in replacement_positions:
                if existing_page_num == page_num and existing_word == word and is_overlapping(existing_rect, rect):
                    break
            else:
                replacement_positions.append((page_num, rect, word, replace))
print(replacement_positions)
# Now perform redactions and insertions
for page_num, rect, word, replace in replacement_positions:
    print(f"processing {word}")
    page = doc[page_num]
    text_instances = page.get_text("dict", clip=rect)["blocks"][0]["lines"][0]["spans"]
    font_size = text_instances[0]["size"]
    origin = text_instances[0]["origin"]
    # Redact the original word
    redaction = page.add_redact_annot(rect+shift_rect)
    page.apply_redactions()
    page.insert_text(origin, replace, fontsize=font_size, fontname="helv", color=(0, 0, 0), )#set_simple=True)
# Save the modified document
doc.save("output35.pdf", garbage=4, deflate=True, clean=True)

```
Also I noticed that output when I print(text_instances):
processing TX
[{'size': 7.591592311859131, 'flags': 0, 'font': 'Helvetica', 'color': 0, 'ascender': 1.0750000476837158, 'descender': -0.29899999499320984, 'text': 'io', 'origin': (252.05101013183594, 388.135986328125), 'bbox': (252.05101013183594, 382.1964111328125, 257.957275390625, 389.78802490234375)}]
As you see it's processing TX and the 'text':'io' it's the last 2 char in word Antonio!

The original:
![image](https://github.com/pymupdf/PyMuPDF/assets/140750834/72e81956-fe3b-4d23-94bf-c4d365eec782)

The replacment:
![image](https://github.com/pymupdf/PyMuPDF/assets/140750834/120b296a-ea16-40cd-9913-6b6f8bbd604d)

**Comment by JorjMcKie** [comment, marked as answer]
Several things may go wrong with this approach:
* Even if old and new text contain the same number of characters **_and_** the same font size is used, this does not mean that the same written length comes out.
* Even if old and new text have the **_same font_** (have they?) - because Helvetica is a proportional font (every character in principle has its own width).
* Obviously, the border wrapping the text has been drawn after the original text's writing. You may need to insert new text in background (overlay=False)  in the hope that this will improve things.

To exactly fill the original rectangle, use the `morph` parameter with a scaling matrix (fixpoint is the origin of course). The matrix should look like `fitz.Matrix(scale, 1)` where the float scale is computed via this little function:
```python
def adjust_matrix(font, bbox, text, fontsize):
    """Compute matrix performing a horizontal scale.

    Args:
        font: Font object
        bbox: bbox of the text to fill
        text: text
        fontsize: fontsize to use
    Returns:
        Horizontal scaling matrix
    """
    tl = font.text_length(text, fontsize=fontsize)
    width = bbox[2] - bbox[0]
    scale = width / tl
    return fitz.Matrix(scale, 1)
```
As parameter `font` you can use `fitz.Font("helv")` in your case.

**Reply by yang-521** [reply]
Can text adjust its vertical height after insertion? After OCR recognizes the image, I use text_length to calculate the text length and insert it. The text length is equal to the length of the bounding box obtained by OCR, but when the text gap of the image is wide, the height of the inserted text will be high. Is there any way to keep the text box consistent with the OCR bounding box as far as possible?

**Reply by JorjMcKie** [reply]
Nothing can be changed **_after_** insertion. You must find a suitable font size / rectangle height yourself based on how OCR has detected text.

**Comment by Muhammadraafat1** [comment]
@JorjMcKie 
Thank you so much for your assistance and for sharing your expertise!

I have last question. there is any method or approach can handle this case:
If we have, replace = {"FLOWER, CALLIE":"JOHN, DOE"} But In another position in pdf we have only "CALLIE" this should replaced by "DOE" but it doesn't. There is any way to handle this? I don't want to write replace = {"FLOWER":"JOHN", "CALLIE":"DOE"}

**Reply by JorjMcKie** [reply]
No, you have to cover all different cases / situations by your code. So you have to be inventive ...
